### ISSUE TITLE
Fix syntax annotation and add syntactic movement support (through alignments/relations) (T062)

### ISSUE THREAD

**Issue by luutuntin** [proycon/flat#138]
Hi Maarten,

I'm exploring FLAT for our annotation project.
Opening [this example](http://proycon.github.io/folia/example.xml) in FLAT,  I can access to a variety of annotation types (see the first attached picture); but when I choose "Syntactic Unit -- syntax-set" and click "Add" button, nothing happens (see the second attached picture).  Do I miss something?

Best,
Alex

_PS_. I did try to create a FoLiA file with two declared annotation types: POS and Syntax, and encounter the same problem: I can add POS annotations but not Syntax ones.

<img width="960" alt="flat_a_variety_of_annotation_types" src="https://user-images.githubusercontent.com/11186240/46271932-dc43ff00-c51c-11e8-9f3c-d63129974ca7.png">
<img width="960" alt="flat_cannot_add_a_syntactic_unit_annotation" src="https://user-images.githubusercontent.com/11186240/46271933-dcdc9580-c51c-11e8-8615-9b160995f4e8.png">

**Comment by proycon** [comment]
Hi Alex,

That sounds like a bug indeed; a new field should appear for syntax annotation when you click the plus button. I'll see if I can reproduce it and fix it (might take a few days).

**Comment by proycon** [comment]
Bug reproduces with javascript error:

```
TypeError: structure[structure_id].spanannotations is undefined[Learn More] flat.js:206:9
forspanannotations
http://mhysa:8080/flat/style/flat.js:206:9
renderparentspanfield
http://mhysa:8080/flat/style/flat.editor.js:510:5
addeditorfield
http://mhysa:8080/flat/style/flat.editor.js:1014:17
editor_oninit/<
http://mhysa:8080/flat/style/flat.editor.js:2121:9
dispatch
http://mhysa:8080/flat/style/jquery-3.1.0.min.js:3:9870
add/q.handle
http://mhysa:8080/flat/style/jquery-3.1.0.min.js:3:7932
```

**Comment by luutuntin** [comment]
Yes, I receive the same javascript errors:

![flat_javascript_errors_edited](https://user-images.githubusercontent.com/11186240/46676213-09904d00-cbae-11e8-973b-10b948e8e51b.png)

**Comment by proycon** [comment]
OK, this should be fixed now in v0.7.14.

**Comment by luutuntin** [comment]
Thank you so much. I just updated to this version and do not encounter this bug.

**Comment by proycon** [comment]
Could you give a specific example of such an annotation perhaps?

**Comment by luutuntin** [comment]
Here is an example from [Syntactic annotation manual](https://www.ling.upenn.edu/~beatrice/annotation-for-audio-aligned-corpora/clauses.html#cp-que-mat) for [AAPCAppE](http://csivc.csi.cuny.edu/aapcappe/):
![syntax_tree](https://user-images.githubusercontent.com/11186240/46809742-5d12be80-cd99-11e8-8cc9-1f085e463eb1.png)
We can see that the wh-noun phrase **WNP-1**, containing **_What_**, is co-indexed with the empty trace element _***T*-1**_ under the noun phrase **NP-PRD**; similarly, the verb _be_ phrase **BEP-2**, containing _**is**_, is co-indexed with the empty element _***-2**_ under the verb _be_ phrase **BEP**.

**Comment by luutuntin** [comment]
There is another bug that is demonstrated in the following example:

**Task:** annotate the syntactic structure of the sentence "_They brought the documents on Tuesday ."_
<img width="946" alt="flat_syntax_01" src="https://user-images.githubusercontent.com/11186240/46844801-8f76e880-cda6-11e8-867c-2c5bb44a84ba.png">

First, assign clause  label **IP-MAT** to the whole sentence:
<img width="948" alt="flat_syntax_02" src="https://user-images.githubusercontent.com/11186240/46844802-8f76e880-cda6-11e8-8de6-99ed9cd936fe.png">
<img width="946" alt="flat_syntax_03" src="https://user-images.githubusercontent.com/11186240/46844803-8f76e880-cda6-11e8-9804-d2123bbf67f4.png">

Next, assign phrase label **NP-SBJ** to "_They_" and set its parent span to **IP-MAT**:
<img width="949" alt="flat_syntax_04" src="https://user-images.githubusercontent.com/11186240/46844804-8f76e880-cda6-11e8-9f65-b07a5ce9399c.png">
<img width="946" alt="flat_syntax_05" src="https://user-images.githubusercontent.com/11186240/46844805-8f76e880-cda6-11e8-8235-9fc27a0f5497.png">
<img width="948" alt="flat_syntax_06" src="https://user-images.githubusercontent.com/11186240/46844806-8f76e880-cda6-11e8-80d7-88e9abdaee57.png">

Next, assign phrase label **VP** to"_brought the documents on Tuesday_" and also set its parent span to **IP-MAT**: 
<img width="947" alt="flat_syntax_07" src="https://user-images.githubusercontent.com/11186240/46844807-8f76e880-cda6-11e8-9cdc-437e40296a2c.png">

After this step, the annotation for "_They_" was unexpectedly excluded from the syntactic annotation:
<img width="946" alt="flat_syntax_08" src="https://user-images.githubusercontent.com/11186240/46844809-8f76e880-cda6-11e8-9921-0d927b34ec4f.png">
<img width="946" alt="flat_syntax_09" src="https://user-images.githubusercontent.com/11186240/46844810-8f76e880-cda6-11e8-8abf-8d3d0143b1dc.png">

**Comment by proycon** [comment]
Thanks for the elaborate report, that indeed looks like another bug (syntactic annotation hasn't been too widely used yet so I'm afraid you're kind of the guinnea pig in this, sorry). I'll investigate and fix it soon!

(I'll get back to you on the syntactic movement issue as well, that may prove challenging in the current setup)

**Comment by luutuntin** [comment]
I really like the FoLiA paradigm and would love to use FLAT for [our project](http://birch.pythonanywhere.com/). For syntactic movement annotation, the simplest solution we are thinking of is just inserting the trace symbols into the text, and then assign syntactic labels to them as to normal tokens. As I understand, we can do this kind of insertion using FLAT, right? Again, thank you so much for your great support.

**Comment by proycon** [comment]
Yes, that simple solution would indeed be a decent workaround and should work right away (after I fix the bug you reported). Perhaps in the underlying FoLiA tokenisation you can also mark such pseudo/trace elements by assigning a special class to the word elements (``<w>``, as you probably know you can make up any vocabulary which is what I'm doing here too with *normal* and *empty*). I worked out your example sentence:

```xml
<s xml:id="s1">
    <w xml:id="w1" class="normal">
        <t>What</t>
        <pos class="WPRO" />
    </w>
    <w xml:id="w2" class="normal">
        <t>is</t>
        <pos class="BEP" />
    </w>
    <w xml:id="w3" class="normal">
        <t>your</t>
        <pos class="PRO$" />
    </w>
    <w xml:id="w4" class="normal">
        <t>name</t>
        <pos class="N" />
    </w>
    <w xml:id="w.bep-2" class="empty">
        <t>*-2</t>
    </w>
    <w xml:id="w.wnp-1" class="empty">
        <t>*T*-1</t>
    </w>
    <w xml:id="w5">
        <t>?</t>
        <pos class="PUNC" />
    </w>
    <syntax>
        <su class="CP-QUE-MAT">
            <su xml:id="s1.WNP-1" class="WNP">
                <wref id="w1" />
            </su>
            <su class="IP-SUB">
                <su xml:id="s1.BEP-2" class="BEP">
                    <wref id="w2" />
                </su>
                <su class="NP-SBJ">
                    <su class="NP-POS">
                        <wref id="w3" />
                    </su>
                    <wref id="w4" />
                </su>
                <su class="VP">
                    <su class="BEP">
                        <wref id="w.bep-2" />
                    </su>
                    <su class="NP-PRD">
                        <wref id="w.wnp-1" />
                    </su>
                </su>
            </su>
            <su class="PUNC">
                <wref id="w5" />
            </su>
        </su>
    </syntax>
</s>
```
I'm not really a fan of using extra words/tokens (``<w>``) and text content (``<t>``) for something that technically is empty, but I admit it would be the most practical workaround.

I'm also thinking what the most elegant representation would be from a FoLiA perspective. I'm not very knowledgeable on syntactic movement, but I guess these trace elements should ideally not be expressed in the tokenisation layer but only as part of the syntax tree? The notion of co-indexed should then also be expressed explicitly rather than conventionally, which could be done with FoLiA's alignments (basically higher-order references). I come up with something like this then:

```xml
<s xml:id="s1">
    <w xml:id="w1">
        <t>What</t>
        <pos class="WPRO" />
    </w>
    <w xml:id="w2">
        <t>is</t>
        <pos class="BEP" />
    </w>
    <w xml:id="w3">
        <t>your</t>
        <pos class="PRO$" />
    </w>
    <w xml:id="w4">
        <t>name</t>
        <pos class="N" />
    </w>
    <w xml:id="w5">
        <t>?</t>
        <pos class="PUNC" />
    </w>
    <syntax>
        <su class="CP-QUE-MAT">
            <su xml:id="s1.WNP-1" class="WNP">
                <wref id="w1" />
            </su>
            <su class="IP-SUB">
                <su xml:id="s1.BEP-2" class="BEP">
                    <wref id="w2" />
                </su>
                <su class="NP-SBJ">
                    <su class="NP-POS">
                        <wref id="w3" />
                    </su>
                    <wref id="w4" />
                </su>
                <su class="VP">
                    <su class="BEP">
                        <alignment class="A-movement">
                            <aref id="s1.BEP-2" type="su"/>
                        </alignment>
                    </su>
                    <su class="NP-PRD">
                        <alignment class="Wh-movement">
                            <aref id="s1.WNP-1" type="su"/>
                        </alignment>
                    </su>
                </su>
            </su>
            <su class="PUNC">
                <wref id="w5" />
            </su>
        </su>
    </syntax>
</s>
```

This looks much cleaner to me than the workaround, though it's currently impossible to do in FLAT and would demand an extension (or we could solve it in a postprocessing conversion script, though less elegant). What do you think?

Thanks for considering FLAT for your project! :) It would indeed be great if it can be applied to your task and is a great test-run for syntax annotation for us as well.

**Comment by luutuntin** [comment]
Yes, the second solution is much cleaner and well demonstrates FoLiA's power. I just wonder what the corresponding syntactic structure looks like. Something likes this?
![syntax_tree_alignment](https://user-images.githubusercontent.com/11186240/46990515-9f0c6f00-d0cf-11e8-8c26-2fbede6b8391.png)

As per our plan, we would love to have a stable tool for intensive syntax annotation next summer (i.e. at the beginning of May, 2019). (Currently, we are working hard on developing the annotation guidelines.) Can this go along with your work on the extension? In any case, thoroughly handling movement annotation is essential for the syntax layer.

Regarding the first solution, I tried adding the annotation to my FoLiA file and received the following tree:
![syntax_tree_insertion](https://user-images.githubusercontent.com/11186240/46991307-6e2e3900-d0d3-11e8-9528-b576306a79fd.png)
We can see that whenever there is a part-of-speech label branch (i.e. a branch whose root node is a part-of-speech label) going out from a syntactic label node, it will not be shown in **Tree Viewer**. Is this a bug?

Another concern is that we don't know which pseudo/trace elements we want to insert until we do the syntax annotation, and therefore we cannot insert them at tokenisation stage. I haven't figured out how to add new tokens using FLAT. Can you give me a hint?

**Comment by proycon** [comment]
I'm debugging the original issue with the dissappearing 'They' and can confirm this indeed goes wrong (just documenting this mostly for my own fixing process, I'll answer the other questions in a separate comment):

> 2018-10-17 20:30:17 - [QUERY ON flat/issue138] ADD su OF test WITH class "IP-MAT" annotator "flat" annotatortype "manual" datetime now confidence NONE FOR SPAN ID issue138.p.1.s.1.w.1 & ID   issue138.p.1.s.1.w.2 & ID issue138.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 & ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FORMAT flat RETURN ancestor-target    

> 2018-10-17 20:38:41 - [QUERY ON flat/issue138] EDIT su ID issue138.text.su.1 WITH class "IP-MAT" annotator "flat" annotatortype "manual" datetime now confidence NONE RESPAN  ID issue138.p.1.s.1.w.2 & ID issue138.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 & ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FOR ID issue138.p.1.s.1.w.2 , ID issue138.p.1.s.1.w.3 , ID issue138.p.1.s.1.w.4 , ID issue138.p.1.s.1.w.5 , ID issue138.p.1.s.1.w.6 , ID issue138.p.1.s.1.w.7 FORMAT flat RETURN ancestor-focus                                                
  
>  2018-10-17 20:38:41 - [QUERY ON flat/issue138] ADD su OF test WITH class "NP-SBJ" annotator "flat" annotatortype "manual" datetime now confidence NONE SPAN  ID issue138.p.1.s.1.w.1 FOR ID i
  ss ue138.text.su.1 FORMAT flat RETURN ancestor-target

>  2018-10-17 20:40:35 - [QUERY ON flat/issue138] EDIT su ID issue138.text.su.1 WITH class "IP-MAT" annotator "flat" annotatortype "manual" datetime now confidence NONE FORMAT flat RETURN ance
  stor-focus
  
>  2018-10-17 20:40:35 - [QUERY ON flat/issue138] ADD su OF test WITH class "VP" annotator "flat" annotatortype "manual" datetime now confidence NONE SPAN  ID issue138.p.1.s.1.w.2 & ID issue13
  8.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 & ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FOR ID issue138.text.su.1 FORMAT flat RETURN ancestor-target

The second query seems the culprit and shouldn't have been formed (the fourth also not).

Additionally, after doing this I end up with an inconsistency in the front end when hovering over the annotation: ``uncaught exception: Error, unable to sort targets, expected 13, got 7, sameparent=issue138.p.1.s.1``. It seems there are duplicate targets because of the nesting.

stack trace:
```
* sort_targets flat.js:755
* getspantext flat.viewer.js:232
```

Todo:
- [ ] Fix getspantext() and sort_targets() to handle nested span elements
- [ ] Fix detection of changes in case of multiple span elements, prevent erroneous respan queries

**Comment by proycon** [comment]
Those respans I said were wrong actually are not wrong! I first respan the parent so it doesn't include the ``wref`` anymore, as that ``wref`` will be covered by the ``su`` of the child and we don't want duplicates. That's why we get an EDIT..RESPAN query followed by an  ADD query every time.

**Comment by proycon** [comment]
After a syntactic unit on the first word, the order is wrong:

```xml
<syntax>
 <su annotator="flat" annotatortype="manual" class="S" datetime="2018-10-18T12:56:21" xml:id="issue138.text.su.1">
   <wref id="issue138.p.1.s.1.w.2" t="brought"/>
   <wref id="issue138.p.1.s.1.w.3" t="the"/>
   <wref id="issue138.p.1.s.1.w.4" t="documents"/>
   <wref id="issue138.p.1.s.1.w.5" t="on"/>
   <wref id="issue138.p.1.s.1.w.6" t="Tuesday"/>
   <wref id="issue138.p.1.s.1.w.7" t="."/>
   <su annotator="flat" annotatortype="manual" class="PRON" datetime="2018-10-18T12:56:21" xml:id="issue138.text.su.2">
       <wref id="issue138.p.1.s.1.w.1" t="They"/>
   </su>
 </su>
</syntax>
```

 (I thought this might explain subissue (c) in the tree visualisation going wrong, but no, that also goes wrong if the order is correct)

Todo:

-  [ ] (d) Insertion point of ADD queries should be computed in case of nested span elements (su)

**Comment by proycon** [comment]
The final respan when clearing all of the remaining parent syntactic unit (which is common when the parent unit is fully covered child syntactic units), doesn't happen is and instead if a kind of no-operation:

```
EDIT su ID issue138.text.su.1 WITH class "S" annotator "flat" annotatortype "manual" datetime now confidence NONE FORMAT flat RETURN ancestor-focus
```

Probably because I didn't allow RESPANs to be empty, but that is valid and necessary here. So, new subissue replacing (b) (which turned out not to be wrong):

- [ ] (e) Allow ``RESPAN NONE`` on parent when inserting children (without deleting the parent span)

**Comment by luutuntin** [comment]
It is really enjoyable to follow your debugging process.

Actually, the visualisation issue I mentioned in my previous post is the same as _(c) Tree visualisation can't deal with heterogenous children (su + wref)_. In the example "What is your name ?", PoS labels include WPRO, BEP, PRO$, N, PUNC ([full reference](https://www.ling.upenn.edu/~beatrice/annotation-for-audio-aligned-corpora/pos.html)).

I also noticed the wrong order you mentioned in your _todo (d)_ post when I experimented with "They brought the documents on Tuesday ."

I just reviewed [possible empty categories](https://www.ling.upenn.edu/~beatrice/annotation-for-audio-aligned-corpora/empty.html#position) in syntax annotation and found that, in addition to traces of movement which can be handled by alignment annotation, there are other empty categories such as empty subjects which cannot be aligned to any available tokens and therefore require the insertion of new tokens. For example, in the following sentence there is an empty expletive subject, **\*exp\***, as the silent counterpart of an existential subject.
![syntax_tree_empty_expletive_subject](https://user-images.githubusercontent.com/11186240/47251308-296f0e80-d401-11e8-9b43-1b00fc0b88fb.png)
If we still want to preserve the original text content layer, should we extend FoLiA's specification so that we can insert additional tokens for this kind of empty categories into the syntax layer (and this may be a desirable feature for other annotations such as implicit semantic roles)? Something likes this:
```
<s xml:id="s.1">
    <w xml:id="s.1.w.1">
        <t>Is@</t>
        <pos class="BEP" />
    </w>
    <w xml:id="s.1.w.2">
        <t>@n't</t>
        <pos class="NEG" />
    </w>
    <w xml:id="s.1.w.3">
        <t>a</t>
        <pos class="D" />
    </w>
    <w xml:id="s.1.w.4">
        <t>whole</t>
        <pos class="ADJ" />
    </w>
    <w xml:id="s.1.w.5">
        <t>lot</t>
        <pos class="N" />
    </w>
    <w xml:id="s.1.w.6">
        <t>left</t>
        <pos class="VAN" />
    </w>
    <w xml:id="s.1.w.7">
        <t>.</t>
        <pos class="PUNC" />
    </w>
    <syntax>        
        <su xml:id="s.1.su.1" class="IP-MAT">
            <su xml:id="s.1.su.2" class="NP-SBJ">
                **<w xml:id="s.1.su.w.1">
                    <t>*exp*</t>
                    <pos class="EX" />
                </w>**
            </su>
            <su xml:id="s.1.su.3" class="VP">
                <su xml:id="s.1.su.4" class="BEP">
                    <wref id="s.1.w.1" />
                </su>
                <su xml:id="s.1.su.5" class="NEG">
                    <wref id="s.1.w.2" />
                </su>
                <su xml:id="s.1.su.6" class="VP">
                    <su xml:id="s.1.su.7" class="NP=LGS">
                        <wref id="s.1.w.3" />
                        <su xml:id="s.1.su.8" class="ADJP">
                            <wref id="s.1.w.4" />
                        </su>
                        <wref id="s.1.w.5" />
                    </su>
                    <wref id="s1.w.6" />
                </su>
            </su>
            <su class="PUNC">
                <wref id="s.1.w.7" />
            </su>
        </su>
    </syntax>
</s>
```

**Comment by proycon** [comment]
Glad the debugging is interesting ;) Sorry it took a while again (doing this in the midst of working on lots of other things too). 

You raise a very interesting point regarding the empty words, we don't really have a good mechanism yet to explicitly accommodate such *ghost words/tokens* (if you have a more linguistically sound term I'd be glad to hear it :) ) so we indeed might need to extend FoLiA there. If you really want to do that in the current FoLiA then I'm thinking of a solution using a set definition for words/tokens that contain a class "empty/implicit/ghost" or however you want to call it, as opposed to a class which I'll "normal".

```xml
<s xml:id="s.1">
    <w xml:id="s.1.w.0" class="ghost">
        <t>*exp*</t>
        <pos class="EX" />
    </w>
    <w xml:id="s.1.w.1" class="normal">
        <t>Is@</t>
        <pos class="BEP" />
    </w>
    <w xml:id="s.1.w.2" class="normal">
        <t>@n't</t>
        <pos class="NEG" />
    </w>
    <w xml:id="s.1.w.3" class="normal">
        <t>a</t>
        <pos class="D" />
    </w>
    <w xml:id="s.1.w.4" class="normal">
        <t>whole</t>
        <pos class="ADJ" />
    </w>
    <w xml:id="s.1.w.5" class="normal">
        <t>lot</t>
        <pos class="N" />
    </w>
    <w xml:id="s.1.w.6" class="normal">
        <t>left</t>
        <pos class="VAN" />
    </w>
    <w xml:id="s.1.w.7" class="normal">
        <t>.</t>
        <pos class="PUNC" />
    </w>
    <syntax>        
        <su xml:id="s.1.su.1" class="IP-MAT">
            <su xml:id="s.1.su.2" class="NP-SBJ">
                <wref id="s.1.w.0" />
            </su>
            <su xml:id="s.1.su.3" class="VP">
                <su xml:id="s.1.su.4" class="BEP">
                    <wref id="s.1.w.1" />
                </su>
                <su xml:id="s.1.su.5" class="NEG">
                    <wref id="s.1.w.2" />
                </su>
                <su xml:id="s.1.su.6" class="VP">
                    <su xml:id="s.1.su.7" class="NP=LGS">
                        <wref id="s.1.w.3" />
                        <su xml:id="s.1.su.8" class="ADJP">
                            <wref id="s.1.w.4" />
                        </su>
                        <wref id="s.1.w.5" />
                    </su>
                    <wref id="s1.w.6" />
                </su>
            </su>
            <su class="PUNC">
                <wref id="s.1.w.7" />
            </su>
        </su>
    </syntax>
</s>
```

But I think you are right that a more explicit notation may be necessary for these kinds of issues, rather than leaving it to the set definition. I like you're suggestion, it's simple enough and currently not allowed so it's on the table as an option. 

We'd have to consider that:
* A word may be referenced multiple times in span annotation (so logically I think the first one could be a ``<w>`` as suggest and the subsequent ones normal ``<wref>``)
*  Span annotation elements are not always strict on order (``<su>`` by definition is though). Having a ``<w>`` appear inside span annotation might leave no clues to where it would be in the original text. Whether that is a problem or not I don't know yet, because they are ghost words/tokens that are not in the original anyway and I do prefer the idea of confining them to their span annotation layer because there may be different kind of ghost words/tokens for different annotation types (you already suggested semantic roles).
* The FoLiA libraries will need some mechanism to determine whether a word/token is a ghost/implicit/empty one or not, as it's common to pass ``Word`` instances to span annotation elements directly which are resolved to wrefs internally.
* For FLAT this raises more challenges as to how actually allow this in the interface.

**Comment by proycon** [comment]
Subissue (c), the treeviewer not showing mixed content (su + wref), is caused by FLAT's representation pulling the two apart. The child su gets seen as an annotation and the wrefs get seen as the target. Unfortunately the ordering information is not readily available at that point so I'll have to devise a solution.

![](https://download.anaproy.nl/issue138c.png)

**Comment by luutuntin** [comment]
> Further debugging:
> 
> Somehow I did end up in a correct state despite the respan error:
> 
> ![Screenshot](https://camo.githubusercontent.com/00da073962c0a2b72fcfe085d24e469bf8f946ef/68747470733a2f2f646f776e6c6f61642e616e6170726f792e6e6c2f69737375653133385f302e6a7067)
> 
> When adding a V under VP the representation seemed fine but the tree visualizer couldn't visualize it properly (perhaps due to one child being a `su` and the others `wref`), so this might be a new subissue:
> 
> * [ ]  (c) Tree visualisation can't deal with heterogenous children (su + wref)
> 
> > 2018-10-18 11:09:35 - [QUERY ON flat/issue138] EDIT su ID issue138.text.su.3 WITH class "VP" annotator "flat" annotatortype "manual" datetime now confidence NONE RESPAN  ID issue138.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 & ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FORMAT flat RETURN ancestor-focus  **(this one is wrong again)**
> 
> > 2018-10-18 11:09:35 - [QUERY ON flat/issue138] ADD su OF test WITH class "V" annotator "flat" annotatortype "manual" datetime now confidence NONE SPAN  ID issue138.p.1.s.1.w.2 FOR ID issue138.text.su.3 FORMAT flat RETURN ancestor-target
> 
> ![Screenshot](https://camo.githubusercontent.com/73eedb30ce51fd11825754b13099081b27b0998f/68747470733a2f2f646f776e6c6f61642e616e6170726f792e6e6c2f69737375653133385f312e6a7067)
> 
> When trying to add a `NP` I again lost a word (due to the first query) which again seems in instance of subissue b, and I ended up with this mess:
> 
> > 2018-10-18 11:12:38 - [QUERY ON flat/issue138] EDIT su ID issue138.text.su.3 WITH class "VP" annotator "flat" annotatortype "manual" datetime now confidence NONE RESPAN  ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FORMAT flat RETURN ancestor-focus  **wrong again!**
> 
> > 2018-10-18 11:12:38 - [QUERY ON flat/issue138] ADD su OF test WITH class "NP" annotator "flat" annotatortype "manual" datetime now confidence NONE SPAN  ID issue138.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 FOR ID issue138.text.su.3 FORMAT flat RETURN ancestor-target
> 
> ![Screenshot](https://camo.githubusercontent.com/0a6eb39513da1c5002f7005c73acfb7214c5dab4/68747470733a2f2f646f776e6c6f61642e616e6170726f792e6e6c2f69737375653133385f322e6a7067)

**Comment by proycon** [comment]
Small update: I haven't forgotten about this but since this issue requires changes in FoLiA and its libraries I'm taking it along in the development of FoLiA v2.0 which is currently in full progress (and which depends on a fair amount of other new stuff too so takes some time).

**Comment by luutuntin** [comment]
I'm sure that you are working hard on certain radical changes. Thank you.

**Comment by proycon** [comment]
Now FoLiA v2 is released, I'm working on FLAT again and making progress regarding this issue. I hadn't commented on this yet though:

> Another note on allowing <w> under span annotation units: in addition to empty categories, we may want to split a word (or join words) in syntax annotation layer, but not in text content layer. For example, in sentence "Lemme go .", we may want to keep the orthographic word "lemme" in text content layer, but split it into "lem@" and "@me" in syntax annotation layer.

That would be possible by introducing a morphology layer on the "lemme" word, with two morphemes, and then link to the morphemes from the syntax layer. FoLiA supports that but it may require some additional work in FLAT still.

**Comment by proycon** [comment]
subissue d) (Insertion point of ADD queries should be computed in case of nested span elements (su)) seems okay now.

**Comment by proycon** [comment]
A new subissue **f** arose, which may be related to **e)** (Allow RESPAN NONE on parent when inserting children (without deleting the parent span) ).

* [ ] **(f)** Left part of tree gets lost when attaching a new right part, parent respanned to cover only the right part, left sibling gets deleted.

This seems an instance of what @luutuntin already reported here:

> After creating IP-MAT for the whole sentence, if I exclude the punctuation token from VP or create NP-SBJ before VP, I will end up loosing VP or NP-SBJ respectively.

FQL queries:

```
USE proycon/issue138 PROCESSOR name "proycon" type manual IN $FLAT_PROCESSOR IN $FOLIADOCSERVE_PROCESSOR EDIT su ID issue138.text.su.1 WITH class "S" datetime now confidence NONE RESPAN NONE FORMAT flat RETURN ancestor-focus

USE proycon/issue138 PROCESSOR name "proycon" type manual IN $FLAT_PROCESSOR IN $FOLIADOCSERVE_PROCESSOR ADD su OF adhoc WITH class "VP" datetime now confidence NONE SPAN  ID issue138.p.1.s.1.w.2 & ID issue138.p.1.s.1.w.3 & ID issue138.p.1.s.1.w.4 & ID issue138.p.1.s.1.w.5 & ID issue138.p.1.s.1.w.6 & ID issue138.p.1.s.1.w.7 FOR ID issue138.text.su.1 FORMAT flat RETURN ancestor-target
```

This seems caused by the ``RESPAN NONE`` (which was also the subject of subissue e), I'll have to make sure that ``RESPAN NONE`` does NOT affect any underlying child elements.

**Comment by proycon** [comment]
Subissue f is fixed now, and I think e is also solved. Subissue c remains still.

**Comment by proycon** [comment]
Tree visualisation (subissue c) is fixed now as well

**Comment by proycon** [comment]
A short summary of dependencies to be implemented for proper syntactic movement support:
* **[done]** ~Add support for hidden words/token and allow them to be targets for references~ #141
* Add support for appending/prepending hidden words in the annotation editor #145
* Implement support for relations #84 

These should be realisable in the immediate short term (I hope), even though #84 is fairly big component.

And what is needed if you want to refer to sub-parts of a token (i.e. morphemes) rather than to a whole token. A work around is to adapt the tokenisation layer (in a preprocessing step prior to FLAT).
* Allowing using morphemes as a target in span annotation #142
* Implement support for viewing and editing of morphemes and phonemes #13  (this would be a big component and not realisible in the immediate short term).

**Comment by proycon** [comment]
@luutuntin Привет! FLAT v0.8.0 has been released a bit over a week ago, implementing a lot of syntax annotation fixes stemming from this issue. As I already mentioned in the previous summary post, I plan to implement relations (#84) for v0.9.0 (aiming for the beginning of June as I have some other priorities in other projects first). I just wanted to check if you guys are still planning on using FLAT for your syntactic movement annotation task, and what your timeline is? Can you also let me know if issue proycon/folia#50 has more or less priority for you?

**Comment by luutuntin** [comment]
We are starting morphological annotation now; therefore, issue [proycon/folia#50](https://github.com/proycon/folia/issues/50) has more priority for us. 
Regarding syntactic annotation, we are developing the guidelines, and will start (rule-based) automatic annotation first, which does not involve using FLAT. I don't think we will start any manual correction of syntactic annotation before July. 
Again, thank you so much.

**Comment by luutuntin** [comment]
In addition, [proycon/flat#134](https://github.com/proycon/flat/issues/134) is also relevant to us, because our annotators of morphological analysis may detect some errors in the transcripts and want to make certain corrections in FLAT, for instance, adding quotes.

**Comment by luutuntin** [comment]
> In addition, [proycon/flat#134](https://github.com/proycon/flat/issues/134) is also relevant to us, because our annotators of morphological analysis may detect some errors in the transcripts and want to make certain corrections in FLAT, for instance, adding quotes.

Sorry for my ignorance. This is solved by [proycon/flat/#145](https://github.com/proycon/flat/issues/145). And I just realized that in [FoLiA documentation](https://folia.readthedocs.io/) the [introduction](https://folia.readthedocs.io/en/latest/hiddentoken_annotation.html#hidden-token-annotation) and [specification](https://folia.readthedocs.io/en/latest/hiddentoken_annotation.html#specification) of hidden token annotation are the same as those of token annotation, which should not be. You might forget to update them.

**Comment by proycon** [comment]
Right, good point, something went wrong there indeed. I'll fix it!

**Comment by proycon** [comment]
I released FLAT v0.9.0 last week, this implements the essentials that should enable syntactic movement annotation, as mentioned in my comment from April 18th. Proper support for alternative annotations is also implemented in the latest release.

I suggest we make a new issue to continue the discussion on additional features needed for syntactic movement (this one is getting rather long and most has been resolved).
